subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -253,7 +253,7 @@ public PathVisitor(MerkleDirectorySnapshotBuilder builder, @Nullable Spec<FileTr
 
         @Override
         public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
-            String fileName = Optional.ofNullable(dir.getFileName()).map(Object::toString).orElse("");
+            String fileName = Optional.ofNullable(dir.getFileName()).orElse(dir).toString();
```

Comment:
What does `toString()` yield for the Unix filesystem root? `"/"` or `""`? I think `/` would be strange as a file name.

Reply:
toString does show `/`, as shown in the test.
That's what I asked [here](https://github.com/gradle/gradle/pull/9592#discussion_r289749800), did I misunderstand something?

Reply:
I was expecting that we use `""` as file name for the root on a Unix filesystem. Does `getRoot().toString()` evaluate to `C:\` or `C:` on Windows? We shouldn't include the `\` there either IMO.

Reply:
`Paths.get(".").toAbsolutePath().getRoot()` displays `C:\` on Windows.
Same for `File.listRoots()[0]`

Reply:
On Windows it is a delimiter, since e.g. `C:\temp` is basically a list of `['C:', 'temp'].join('\')`.
But on Linux it actually starts with `/`, so I would rather keep it - but don't mind either way.
Please let me know which makes more sense to you.
cc: @lptr

Reply:
For linux, a relative path starting from the root would then be `['/', 'usr', 'bin']`. Joining that on `/` yields `//usr/bin`. That is why I would like to use the empty String for the root on Unix, so that `['', 'usr', 'bin'].join('/') == '/usr/bin'`.

I wonder where it actually matters how we call the root directory. What happens for a copy task, which takes the name of the directory into account?

Reply:
It seems to me it's mostly used for blacklist matches (both should be safe) and https://github.com/gradle/gradle/blob/3903e8fbbc8893fe16c12709dff0f79850425b81/subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java#L309 (where I'm afraid none of the above will work)

Reply:
We don't check the root directory: https://github.com/gradle/gradle/blob/3903e8fbbc8893fe16c12709dff0f79850425b81/subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java#L82

Reply:
I saw this, but we're still using the `name` inside the block, if `isRoot` is true, right?

Reply:
Yes, we do. When is `getRelativePath` used for the root directory?

Reply:
if it's called, it will use the stored `name` to reconstruct the path. Dunno if it happens for root folders or not.

Reply:
As discussed, we'll leave empty String for the name to avoid OS complications - it's not really used in the case of root folders anyway